### PR TITLE
Fix an error during updating cards which are dashcards

### LINKED ISSUE #31766: False "Cannot read properties" error appears sometimes when saving a question
## Context
Not something that I can reproduce every time, but I am seeing this error fairly frequent (at least once a day) when I edit questions in a dashboard and try to save: `Cannot read properties of undefined (reading 'id')`. Despite the error, it _does_ save as there is no error in the API call nor js console. [Slack context](https://metaboat.slack.com/archives/C01LQQ2UW03/p1687301192261419).

## Steps to Reproduce
If you duplicate the [Self-Service Business](https://stats.metabase.com/dashboard/357) dashboard into your personal collection, and then try to edit each card (changing the name or the sql), and save, you should be able to hit the problem with at least one of the cards. I click thru each card from the dashboard, click on question link if not showing the original question, make edits, then save, and click on the “Back to <dashboard>” link next to question title to go back to the dashboard to repeat the process.

Image

### PR DESCRIPTION
**Pull request by uladzimirdev**
Closes https://github.com/metabase/metabase/issues/31766

### Description

Dashboards with question and text/link cards corrupted dashcards data during saving updates. It turned out that UI cards without a reference to actual card weren't filter during data restoring.

### How to verify

1. Create a dashboard with a question and text card, save it. 
2. Edit text card on the dashboard and save it again.
3. Drill down to the question from dashboard, change visualization type, click save and replace.
There shouldn't be an error like in the issue and question should be saved.

### Demo

https://www.loom.com/share/a01a50f859074500971b0bdb2c70b2e3

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

### CODE REVIEW

File: frontend/src/metabase/dashboard/actions/tabs.ts [hunk 1]
```diff
@@ -293,9 +293,14 @@ export const tabsReducer = createReducer<DashboardState>(
         }
 
         // 1. Replace temporary with real dashcard ids
-        const prevCards = prevDash.ordered_cards.filter(
-          id => !state.dashcards[id].isRemoved,
-        );
+        const prevCards = prevDash.ordered_cards.filter(id => {
+          const dashcard = state.dashcards[id];
```
Comment (uladzimirdev) [inline frontend/src/metabase/dashboard/actions/tabs.ts:297]: actual fix is here

File: frontend/src/metabase/dashboard/actions/tabs.ts [hunk 2]
```diff
@@ -293,9 +293,14 @@ export const tabsReducer = createReducer<DashboardState>(
...
+          const isNotRemoved = !dashcard.isRemoved;
+          const isQuestionCard = dashcard.card_id != null;
```
Comment (uladzimirdev) [inline frontend/src/metabase/dashboard/actions/tabs.ts:299]: dashcards without card_id are visual cards like heading or text, we shouldn't set `dashcardData` for them

File: frontend/src/metabase/dashboard/actions/tabs.ts [hunk 3]
```diff
@@ -293,9 +293,14 @@ export const tabsReducer = createReducer<DashboardState>(
...
+
+          return isNotRemoved && isQuestionCard;
+        });
+
         prevCards.forEach((oldId, index) => {
           state.dashcardData[newCards[index].id] = state.dashcardData[oldId];
```
Comment (kamilmielnik) [inline frontend/src/metabase/dashboard/actions/tabs.ts:305]: :x:

I'm pretty sure that source of the problems is here: there is an assumption that `index` will point to the same thing in `prevCards` and `newCards` arrays.

I followed reproduction steps from PR description, and then did this:

1. Refreshed the page
2. Edit dashboard
3. Add new chart to the dashboard
4. Drag & drop the chart to be the first chart in the dashboard
5. Save it

The new chart stays in loading state forever. This bug is not present in `master`.

https://github.com/metabase/metabase/assets/6830683/26a4f942-83ff-4b02-b8d3-61888e8150af

**Review by kamilmielnik — DISMISSED** [review]
Good job on finding reproduction steps for this issue :+1: 

However I found that this PR introduces/exposes a different/deeper problem.

File: e2e/test/scenarios/dashboard/reproductions/34626-corrupted-dashcards-after-save.cy.spec.js [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+import { SAMPLE_DATABASE } from "e2e/support/cypress_sample_database";
+
+const { ORDERS_ID } = SAMPLE_DATABASE;
+
+describe("issue 34626", () => {
+  beforeEach(() => {
+    restore();
+    cy.signInAsAdmin();
+  });
+  it("should not corrupt dashboard data (metabase#34626)", () => {
```
[the remark this answers is not in the document]
Comment (uladzimirdev) [reply e2e/test/scenarios/dashboard/reproductions/34626-corrupted-dashcards-after-save.cy.spec.js]: good point, fixed

File: frontend/src/metabase/dashboard/actions/tabs.ts (hunk 3 above)
Comment (uladzimirdev) [reply frontend/src/metabase/dashboard/actions/tabs.ts]: nice catch, I was afraid of breaking something 100%, please check cd6fe55fef733f144a19a496ac28c3068a5b1613

Comment (uladzimirdev) [reply frontend/src/metabase/dashboard/actions/tabs.ts]: This error about reading "id" is not from here, but from dashboard reducers. Try the same steps to get an error on master branch but in Firefox - you'll see it

**Review by EmmadUsmani — APPROVED** [review]
Changes related to tabs look good to me! I left one small suggestion

File: e2e/test/scenarios/dashboard/reproductions/31766-corrupted-dashcards-after-save.cy.spec.js [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    cy.log("Update viz settings");
+
+    cy.findByTestId("view-footer")
+      .findByRole("button", { name: "Visualization" })
+      .click();
+    cy.findByTestId("Detail-button").click();
+
+    saveUpdatedQuestion();
+
+    modal().should("not.exist");
```
Comment (kamilmielnik) [inline e2e/test/scenarios/dashboard/reproductions/31766-corrupted-dashcards-after-save.cy.spec.js:69]: Just from the test code it's not clear why the modal could not disappear.
I'd add an explicit assertion checking that the error message is not there.

**Review by kamilmielnik — APPROVED** [review]
LGTM :+1:

File: frontend/src/metabase/dashboard/reducers.js [hunk 6]
```diff
@@ -271,6 +271,9 @@ const isNavigatingBackToDashboard = handleActions(
   INITIAL_DASHBOARD_STATE.isNavigatingBackToDashboard,
 );
 
+/**
+ * There is another place where the state is updated - `tabsReducer`
```
[the remark this answers is not in the document]
Comment (kamilmielnik) [reply frontend/src/metabase/dashboard/reducers.js:275]: I'd expect this comment to be as close to the reducer as possible.
Suggested change reads better though :+1:

File: e2e/test/scenarios/dashboard/reproductions/31766-corrupted-dashcards-after-save.cy.spec.js (hunk 5 above)
Comment (uladzimirdev) [reply e2e/test/scenarios/dashboard/reproductions/31766-corrupted-dashcards-after-save.cy.spec.js]: I hope this one is better cbc2f2e9140d49b8b9b73c7f511abb402f2c7db2

Comment (kamilmielnik) [reply e2e/test/scenarios/dashboard/reproductions/31766-corrupted-dashcards-after-save.cy.spec.js]: Nice :+1:

**Pull request merged**